### DISCUSSION TITLE
Scope of lookup for annotation to replaceable model?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#10489] [Q&A] [answered]
When opening some models, e.g. `Modelica.Fluid.Vessels.ClosedVolume`, OMEdit started emitting Scripting Errors:
* Start OMEdit
* Open `Modelica.Fluid.Vessels.ClosedVolume`
* Ignore the Translation warning about a missing `system`
* Ignore a Scripting Error that "Medium is partial, name lookup is not allowed in partial classes".
* Notice a Scripting Error "Modelica 4.0.0+maint.om/Fluid/Vessels.mo:231:73-231:96:writable] Error: Variable use_HeatTransfer not found in scope HeatTransfer."

The relevant bit in the MSL 4 code is
```modelica
       parameter Boolean use_HeatTransfer = false
        "= true to use the HeatTransfer model"
            annotation (Dialog(tab="Assumptions", group="Heat transfer"));
        replaceable model HeatTransfer =
            Modelica.Fluid.Vessels.BaseClasses.HeatTransfer.IdealHeatTransfer
          constrainedby
        Modelica.Fluid.Vessels.BaseClasses.HeatTransfer.PartialVesselHeatTransfer
        "Wall heat transfer"
            annotation (Dialog(tab="Assumptions", group="Heat transfer",enable=use_HeatTransfer),choicesAllMatching=true);
```
in the annotation of `HeatTransfer`, argument `enable`.

I'm not sure why the search for `use_HeatTransfer` is restricted to `HeatTransfer`. First I thought that it is allowed to use external variables in annotations, but then I could not find a definitive answer in the specification, and I could not disentangle the lookup rules between `replaceable`, `annotation`, type definitions vs. "normal" modifications. 
The only thing I could find was in https://specification.modelica.org/master/scoping-name-lookup-and-flattening.html#simple-name-lookup that 
> Reference to variables successfully looked up in an enclosing class is only allowed for variables declared as constant

but I'm not sure if that is applicable here. 
I thought that is is not really useful if one cannot use model parameters in annotations, but then I realised that the difference of the present problem compared to more common annotation usage is that we are effectively annotation not an _instance_ declaration, but a _type_ definition. (maybe?)

So, at this point I'm confused and not sure if could be an OMEdit bug, an MSL issue, or a problem with the specification, and I though I'd ask in the discussions, first. 

 - OpenModelica v1.21.0-dev-318-gd7cb3a8558, Windows 10, Modelica 4.0.0+maint.om

**Comment by perost** [comment, marked as answer]
Fixed in #10491.

The issue was that we started the lookup from `HeatTransfer`, which didn't work when `HeatTransfer` was used as the type of the component `heatTransfer`. The scope of an annotation is always the same scope as where it's defined, i.e. in the scope where `HeatTransfer` is defined rather than `HeatTransfer` itself.

**Reply by bilderbuchi** [reply]
Thanks for the crazy quick fix, and the explanation!
> The scope of an annotation is always the same scope as where it's defined, i.e. in the scope where HeatTransfer is defined rather than HeatTransfer itself.

That's good to get confirmation for, makes sense!
